### PR TITLE
Update tools used

### PR DESCRIPTION
**Pull request by AlinderS**
Replace `black` with the Ruff linter.
Remove `py-heat-magic` as it has not been updated in several years and I was unable to install it on an updated system.
Update the format of ruff.toml to the modern standard.
Change the Ruff rule selecting from `"ALL"` to a select group. This is because the code contains multiple deliberately bad or inconsistent examples as well as examples of code that is more explicit than compact. Putting up exceptions for these examples would be distracting in the presentations and would be too much work to maintain as the rules are frequently updated.

Update all pre-commit hooks to latest versions.

Fix some issues found by the newer versions of linters and formatters. Primarily instances of [B007](https://docs.astral.sh/ruff/rules/unused-loop-control-variable/) but also one instance of [UP034](https://docs.astral.sh/ruff/rules/extraneous-parentheses/).
Fix some typos discovered by the new version of `codespell`.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -9,29 +9,25 @@ repos:
...
-
 -   repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: "v0.1.5"
+    rev: "v0.6.2"
     hooks:
       - id: ruff
         args: ["--fix", "--show-fixes"]
         exclude: 4-tools/is_this_pep8.py
+      - id: ruff-format
+        exclude: 4-tools/is_this_pep8.py
```
Comment (eerovaher) [inline .pre-commit-config.yaml:19]: A better idea would be to exclude this file once in `ruff.toml`, not twice in the `pre-commit` configuration.

File: 4-tools/manual.ipynb [hunk 2]
```diff
@@ -25,7 +25,7 @@
     "\n",
     "[The Python Style Guide](https://www.python.org/dev/peps/pep-0008/), commonly referred to as PEP 8, was already mentioned in the first lesson. Here we introduce tools that allow you to check if your code is PEP 8 compliant. \n",
     "\n",
-    "## Ruff\n",
+    "## Ruff linter\n",
```
Comment (eerovaher) [inline 4-tools/manual.ipynb:28]: Better to have a single section for Ruff than separate sections for Ruff linter and Ruff formatter.

File: 4-tools/manual.ipynb [hunk 3]
```diff
@@ -494,7 +451,7 @@
    "source": [
     "from tqdm import trange\n",
     "\n",
-    "for i in trange(5):\n",
+    "for _i in trange(5):\n",
```
Comment (eerovaher) [inline 4-tools/manual.ipynb:454]:
```suggestion
    "for _ in trange(5):\n",
```

File: 4-tools/manual.ipynb [hunk 4]
```diff
@@ -557,11 +514,11 @@
     "from tqdm.notebook import tqdm, trange\n",
     "\n",
     "for dataset in tqdm((\"a\", \"b\", \"c\"), desc=\"Working hard\"):\n",
-    "    for step in trange(4, desc=\"Processing dataset \" + dataset):\n",
+    "    for _step in trange(4, desc=\"Processing dataset \" + dataset):\n",
```
Comment (eerovaher) [inline 4-tools/manual.ipynb:517]:
```suggestion
    "    for _ in trange(4, desc=\"Processing dataset \" + dataset):\n",
```

File: 4-tools/manual.ipynb [hunk 5]
```diff
@@ -557,11 +514,11 @@
...
     "        sleep(0.25)\n",
     "\n",
     "for dataset in tqdm((\"x\", \"y\", \"z\"), desc=\"Working some more\"):\n",
-    "    for step in trange(4, desc=\"Processing dataset \" + dataset, leave=False):\n",
+    "    for _step in trange(4, desc=\"Processing dataset \" + dataset, leave=False):\n",
```
Comment (eerovaher) [inline 4-tools/manual.ipynb:521]:
```suggestion
    "    for _ in trange(4, desc=\"Processing dataset \" + dataset, leave=False):\n",
```

File: 4-tools/presentation.ipynb [hunk 6]
```diff
@@ -571,7 +518,7 @@
     "    a = \"hello {}\".format(world)\n",
     "    f = rf\"hello {world}\"\n",
     "if this and that:\n",
-    "    y = \"hello \" \"world\"  # FIXME: https://github.com/psf/black/issues/26\n",
+    "    y = \"hello \" \"world\"  # Comment\n",
```
Comment (eerovaher) [inline 4-tools/presentation.ipynb:521]: Does Ruff not merge the string literals?

File: 4-tools/presentation.ipynb [hunk 7]
```diff
@@ -782,9 +729,9 @@
     "\n",
     "for t in tqdm((0.5, 1, 0.5, 1)):\n",
     "    sleep(t)\n",
-    "for i in trange(5):\n",
+    "for _i in trange(5):\n",
```
Comment (eerovaher) [inline 4-tools/presentation.ipynb:732]:
```suggestion
    "for _ in trange(5):\n",
```

File: 4-tools/presentation.ipynb [hunk 8]
```diff
@@ -782,9 +729,9 @@
...
     "    sleep(0.5)\n",
-    "for i in trange(3, desc=\"Doing important work: \"):\n",
+    "for _i in trange(3, desc=\"Doing important work: \"):\n",
```
Comment (eerovaher) [inline 4-tools/presentation.ipynb:734]:
```suggestion
    "for _ in trange(3, desc=\"Doing important work: \"):\n",
```

File: 4-tools/presentation.ipynb [hunk 9]
```diff
@@ -817,9 +764,9 @@
     "\n",
     "for t in tqdm((0.5, 1, 0.5, 1)):\n",
     "    sleep(t)\n",
-    "for i in trange(5):\n",
+    "for _i in trange(5):\n",
```
Comment (eerovaher) [inline 4-tools/presentation.ipynb:767]:
```suggestion
    "for _ in trange(5):\n",
```

File: 4-tools/presentation.ipynb [hunk 10]
```diff
@@ -817,9 +764,9 @@
...
     "    sleep(0.5)\n",
-    "for i in trange(3, desc=\"Doing important work: \"):\n",
+    "for _i in trange(3, desc=\"Doing important work: \"):\n",
```
Comment (eerovaher) [inline 4-tools/presentation.ipynb:769]:
```suggestion
    "for _ in trange(3, desc=\"Doing important work: \"):\n",
```

File: ruff.toml [hunk 11]
```diff
@@ -1,13 +1,28 @@
+line-length = 90
```
Comment (eerovaher) [inline ruff.toml:1]: Why a non-default value?

File: ruff.toml [hunk 12]
```diff
@@ -1,13 +1,28 @@
...
+indent-width = 4
+
+[format]
+# Use same settings as Black.
+quote-style = "double"
+indent-style = "space"
+docstring-code-format = true
+skip-magic-trailing-comma = false
+line-ending = "auto"
```
Comment (eerovaher) [inline ruff.toml:10]: I would only keep non-default values (if there are any).

File: ruff.toml [hunk 13]
```diff
@@ -1,13 +1,28 @@
...
-select = ["ALL"]
-ignore = [
-    "INP001",   # implicit-namespace-package
-    "COM812",   # TrailingCommaMissing
-    "COM819",   # TrailingCommaProhibited
+select = [
+    # pycodestyle
+    "E",
+    # pyupgrade
+    "UP",
```
Comment (eerovaher) [inline ruff.toml:17]: [What Python version are you targeting?](https://docs.astral.sh/ruff/settings/#target-version)

File: ruff.toml [hunk 14]
```diff
@@ -1,13 +1,28 @@
...
+    "NPY",
 ]
 
-[pydocstyle]
+[lint.pydocstyle]
 convention = "numpy"
 
-[flake8-annotations]
+[lint.flake8-annotations]
 ignore-fully-untyped = true # Codebase has no types
```
Comment (eerovaher) [inline ruff.toml:28]: There's nothing to configure here if you are no longer enforcing these rules.

**Review by eerovaher — COMMENTED** [review]
The updates are reasonable overall, but there are some details that can be improved.

File: ruff.toml (hunk 11 above)
Comment (AlinderS) [reply ruff.toml:1]: We have several lines that are ~90 characters long that are all comments explaining what the code is doing, particularly in the NumPy session.
I do not feel like editing these to be a few characters shorter as they do a good job as they are and I don't think we need to aggressively try to keep the lines short in this material.

**Pull request merged**